docs/proposals/project-user-resource-quotas.md

```diff
@@ -0,0 +1,209 @@
+
+## Implementation
+
+### CRD Changes
+
+Add a new flexible ResourceQuota CRD which will hold the desired quota and current consumption of the quota.
+
+```go
+type ResourceQuota struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   ResourceQuotaSpec   `json:"spec,omitempty"`
+    Status ResourceQuotaStatus `json:"status,omitempty"`
+}
+
+type ResourceQuotaSpec struct {
+    // QuotaSubject describes the object for which the quota is applied to.
+    QuotaSubject *QuotaSubject `json:"quotaSubject"`
+    // ResourceQuotas is a map of maximum resource quotas per resource
+    ResourceQuotas corev1.ResourceList `json:"resourceQuotas"`
+}
+
+type QuotaSubject struct {
+    // Name of the quota subject
+    Name string `json:"name"`
+}
+
+type ResourceQuotaStatus struct {
+    // ResourceConsumption is map which holds the current usage of resources for all seeds
+    ResourceConsumption corev1.ResourceList `json:"resourceConsumption"`
+    // LocalConsumption is a map of the current usage of resources for the local seed
+    LocalConsumption corev1.ResourceList `json:"localConsumption"`
+
+}
+```
+
+The ResourceQuota is a master-cluster resource. It will need to be synced to seed clusters to allow webhooks access.
+As we have a distributed platform over multiple seeds, each seed will have to report its usage in the status. Then the master quota controller
+can recalculate the total resource consumption and re-sync the quotas on the seeds.
+
+### How to sync quota info from seed clusters and master cluster and back
+
+We would need a controller for ResourceQuotas in the master cluster: `master-resource-quota-controller` - MRQC
+
+The MRQC would watch master cluster ResourceQuotas. These are the ones that are created through the API, and maybe one day through `kubectl`.
+The role of the MRQC would be to reconcile the master ResourceQuotas to the seed clusters.
+It also needs to watch seed cluster ResourceQuotas for changes in LocalConsumption, and update the master ResourceQuota ResourceConsumption based on it.
+
+### How to ensure that we get proper resource usage and that it stays in sync?
+
+The obvious way seems to get the amount of resources requested when a KKP MachineDeployment is being created, and then check against
+the project quota and fill it up accordingly. But this would only work through the API, and if there are some issues
+we could have resource quotas which are out of sync.
+
+The safest way would be to check the Node capacity on user clusters, but there are some issues with that as well, as the Nodes get created
+based on the MachineDeployments. So there could be races in between when the Node is created(which can take some time) and 
+when we check if user can create a MachineDeployment.
+
+So we need to focus on the MachineDeployments and control the quota through them. The nodes get created anyway based 
```

Comment:
This chapter talks about MachineDeployments, but what about standalone Machine objects? Shouldn't we focus on those?